File: src/lib/loginRemember.ts
const KEY = 'sgb_login_remember_v1';

export type LoginRememberRole = 'guard' | 'admin' | 'resident';

export type LoginRememberV1 = {
  v: 1;
  societyId?: string;
  role?: LoginRememberRole;
  flatId?: string;
  /** Last used mobile / phone (digits), for resident or guard OTP / password login. */
  phone?: string;
};

function normalizePhone(phone: string | undefined): string | undefined {
  if (!phone) return undefined;
  const digits = phone.replace(/\D/g, '').slice(-10);
  return digits.length >= 10 ? digits : phone.replace(/\D/g, '') || undefined;
}

export function readLoginRemember(): LoginRememberV1 | null {
  try {
    const raw = localStorage.getItem(KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw) as LoginRememberV1;
    if (parsed?.v !== 1) return null;
    return parsed;
  } catch {
    return null;
  }
}

/** Merge-patch remembered login fields (flat + mobile stay editable in the UI). */
export function writeLoginRemember(patch: Partial<Omit<LoginRememberV1, 'v'>>): void {
  try {
    const prev = readLoginRemember() ?? { v: 1 as const };
    const next: LoginRememberV1 = { ...prev, v: 1 };

    if ('societyId' in patch) {
      next.societyId = patch.societyId?.trim() || undefined;
    }
    if ('role' in patch) {
      next.role = patch.role || undefined;
    }
    if ('flatId' in patch) {
      next.flatId = patch.flatId?.trim() || undefined;
    }
    if ('phone' in patch) {
      next.phone = normalizePhone(patch.phone);
    }

    localStorage.setItem(KEY, JSON.stringify(next));
  } catch {
    /* quota / private mode */
  }
}

File: mobile/lib/services/login_remember.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

/// Remembers last login society / role / flat / mobile for autofill (fields stay editable).
class LoginRemember {
  LoginRemember._();

  static const _key = 'sgb_login_remember_v1';

  static Future<LoginRememberData?> read() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_key);
    if (raw == null) return null;
    try {
      final map = jsonDecode(raw) as Map<String, dynamic>;
      if (map['v'] != 1) return null;
      return LoginRememberData(
        societyId: map['societyId'] as String?,
        role: map['role'] as String?,
        flatId: map['flatId'] as String?,
        phone: map['phone'] as String?,
      );
    } catch (_) {
      return null;
    }
  }

  static Future<void> write({
    String? societyId,
    String? role,
    String? flatId,
    String? phone,
    bool clearRole = false,
    bool clearFlat = false,
    bool clearPhone = false,
  }) async {
    final prev = await read();
    final next = LoginRememberData(
      societyId: societyId ?? prev?.societyId,
      role: clearRole ? null : (role ?? prev?.role),
      flatId: clearFlat ? null : (flatId ?? prev?.flatId),
      phone: clearPhone ? null : _normalizePhone(phone ?? prev?.phone),
    );
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _key,
      jsonEncode({
        'v': 1,
        if (next.societyId != null) 'societyId': next.societyId,
        if (next.role != null) 'role': next.role,
        if (next.flatId != null) 'flatId': next.flatId,
        if (next.phone != null) 'phone': next.phone,
      }),
    );
  }

  static String? _normalizePhone(String? phone) {
    if (phone == null || phone.trim().isEmpty) return null;
    final digits = phone.replaceAll(RegExp(r'\D'), '');
    if (digits.length >= 10) return digits.substring(digits.length - 10);
    return digits.isEmpty ? null : digits;
  }
}

class LoginRememberData {
  const LoginRememberData({
    this.societyId,
    this.role,
    this.flatId,
    this.phone,
  });

  final String? societyId;
  final String? role;
  final String? flatId;
  final String? phone;
}
